### PR TITLE
IGNITE-16354 Fix memory leak in RelOptCluster during mapping

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/MappingQueryContext.java [hunk 1]
```diff
@@ -45,4 +53,16 @@ public UUID localNodeId() {
     public AffinityTopologyVersion topologyVersion() {
         return topVer;
     }
+
+    /** Creates a cluster. */
+    RelOptCluster cluster() {
+        if (cluster == null) {
+            cluster = RelOptCluster.create(Commons.emptyCluster().getPlanner(), Commons.emptyCluster().getRexBuilder());
```
Comment (AMashenkov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/MappingQueryContext.java:60]: Seems, VolcanoPlanner can cache schemas (see VolcanoPlanner.registerSchema).
Would you mind to fix the planner as well?

We have similar fix in Ignite-3 (commit d87f1e18) with the method stub and tests looks fine.
But I've tried to apply the same to your PR, but it lead to 2 tests (in PlannerTest) failure with no obvious reason.

Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/MappingQueryContext.java:60]: The planner shouldn't be used by the empty cluster at all, there are assertions to prevent this (I've looked at the fix to ignite-3, and looks like the same check was implemented but in a different way, by prohibiting registerSchema invocation). I can add the same check to the planner as well, but now from my point of view, it looks redundant.
The current fix prohibits not only planning but also mapping (mapping use metadataQuerySupplier, I'm not sure mapping uses planner somehow). 
Also, I think the fix for ignite-3 is not complete, we should attach fragments to the empty cluster again before storing it   to `executionPlan` variable (see https://github.com/apache/ignite/pull/9758/files#diff-07fea6ecf0f238df45e8d6fdffabf83cf9a5fdbd90b31f897868a282c0694f0eR99), otherwise cluster with cached metadata can't be garbage collected (referenced from query cache -> query template -> execution plan -> fragments -> cluster)

Comment (AMashenkov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/MappingQueryContext.java:60]: >  I can add the same check to the planner as well, but now from my point of view, it looks redundant.
I've tried to add the check to your PR code and found 2 tests failed because planner registers schema in empty cluster.
(empty Cluster -> VolcanoPlanner -> registeredSchemas).

> Also, I think the fix for ignite-3 is not complete, we should attach fragments to the empty cluster again before storing it to executionPlan variable

Looks like you are right.

Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/MappingQueryContext.java:60]: > I've tried to add the check to your PR code and found 2 tests failed because planner registers schema in empty cluster. (empty Cluster -> VolcanoPlanner -> registeredSchemas).

Hmm, when I add assert to the `registerSchema` there are a lot of tests failed, because this method is invoked in TableScan rel creation, and this rel is created even when attached to the empty cluster (while cloning rels, while creating it on remote nodes from JSON, etc). I override this method for the empty planner as no-op, please have a look again.

**Review by ivandasch — APPROVED** [review]

**Review by AMashenkov — APPROVED** [review]

**Comment by alex-plekhanov** [comment]
Merged to sql-calcite branch
